packages/back-end/src/models/DataSourceModel.ts

```diff
@@ -336,6 +343,71 @@ export async function updateDataSource(
     return;
   }
 
+  if (
+    isDataSourceType<GrowthbookClickhouseDataSource>(
+      datasource,
+      "growthbook_clickhouse"
+    ) &&
+    isPartialWithMaterializedColumns(updates)
+  ) {
+    updates.settings!.materializedColumns = sanitizeMaterializedColumns(
+      updates.settings!.materializedColumns!
+    );
+    const finalColumns = updates.settings!.materializedColumns!;
+
+    const originalColumns = sanitizeMaterializedColumns(
+      datasource.settings.materializedColumns || []
+    );
+    const newColumnMap = Object.fromEntries(
+      finalColumns.map((col) => [col.sourceField, col])
+    );
+    const originalColumnMap = Object.fromEntries(
+      originalColumns.map((col) => [col.sourceField, col])
+    );
+    const columnsToDelete: MaterializedColumn[] = [],
+      columnsToRename: { from: string; to: string }[] = [];
+
+    originalColumns.forEach((col) => {
+      if (
+        !Object.prototype.hasOwnProperty.call(newColumnMap, col.sourceField)
```

Comment:
This syntax is a little awkward and unnecessarily complex.  Either just use the `key in obj` syntax or make it an actual Map object so you get the `.has` method.